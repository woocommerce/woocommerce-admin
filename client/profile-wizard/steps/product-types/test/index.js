/**
 * External dependencies
 */
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import { ProductTypes } from '../';

const testProps = {
	countryCode: 'US',
	invalidateResolution: jest.fn(),
	isProductTypesRequesting: false,
	productTypes: {
		paidProduct: {
			description: 'Paid product type',
			label: 'Paid product',
			more_url: 'https://woocommerce.com/paid-product',
			product: 100,
			slug: 'paid-product',
			yearly_price: 120,
		},
		freeProduct: {
			label: 'Free product',
		},
	},
};

describe( 'ProductTypes', () => {
	afterEach( () => {
		window.wcAdminFeatures.subscriptions = false;
	} );

	test( 'should render product types', () => {
		const { container } = render( <ProductTypes { ...testProps } /> );
		expect( container ).toMatchSnapshot();
	} );

	test( 'should show annual prices on toggle', () => {
		const { container } = render( <ProductTypes { ...testProps } /> );

		const toggle = screen.getByLabelText( 'Display monthly prices', {
			selector: 'input',
		} );

		userEvent.click( toggle );

		expect( container ).toMatchSnapshot();
	} );

	test( 'should validate on continue', async () => {
		const mockCreateNotice = jest.fn();
		const mockGoToNextStep = jest.fn();
		const mockUpdateProfileItems = jest.fn().mockResolvedValue();

		render(
			<ProductTypes
				createNotice={ mockCreateNotice }
				goToNextStep={ mockGoToNextStep }
				updateProfileItems={ mockUpdateProfileItems }
				{ ...testProps }
			/>
		);

		const continueButton = screen.getByText( 'Continue', {
			selector: 'button',
		} );
		const productType = screen.getByText( 'Free product', {
			selector: 'label',
		} );

		// Validation should fail since no product types are selected.
		userEvent.click( continueButton );
		await waitFor( () => {
			expect( mockGoToNextStep ).not.toHaveBeenCalled();
		} );
		expect( mockUpdateProfileItems ).not.toHaveBeenCalled();

		// Click on a product type to pass validation.
		userEvent.click( productType );
		userEvent.click( continueButton );
		await waitFor( () => {
			expect( mockUpdateProfileItems ).toHaveBeenCalled();
		} );
		expect( mockGoToNextStep ).toHaveBeenCalled();
	} );
	test( 'should show a warning message at the bottom of the step', () => {
		const productTypes = {
			subscriptions: {
				label: 'Subscriptions',
			},
		};
		window.wcAdminFeatures.subscriptions = true;

		render(
			<ProductTypes { ...testProps } productTypes={ productTypes } />
		);

		const subscription = screen.getByText( 'Subscriptions', {
			selector: 'label',
		} );
		userEvent.click( subscription );

		expect(
			screen.queryByText(
				'The following extensions will be added to your site for free: WooCommerce Payments. An account is required to use this feature.'
			)
		).toBeInTheDocument();
	} );
	test( 'should show the warning message only for US stores', () => {
		const productTypes = {
			subscriptions: {
				label: 'Subscriptions',
			},
		};
		const countryCode = 'FR';
		window.wcAdminFeatures.subscriptions = true;

		render(
			<ProductTypes
				{ ...testProps }
				productTypes={ productTypes }
				countryCode={ countryCode }
			/>
		);

		const subscription = screen.getByText( 'Subscriptions', {
			selector: 'label',
		} );
		userEvent.click( subscription );

		expect(
			screen.queryByText(
				'The following extensions will be added to your site for free: WooCommerce Payments. An account is required to use this feature.'
			)
		).not.toBeInTheDocument();
	} );
} );
