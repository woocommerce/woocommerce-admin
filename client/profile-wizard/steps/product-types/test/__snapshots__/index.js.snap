// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTypes should render product types 1`] = `
<div>
  <div
    class="woocommerce-profile-wizard__product-types"
  >
    <div
      class="woocommerce-profile-wizard__step-header"
    >
      <h2
        class="components-truncate components-text css-7a2wz9-View-Text-sx-Base em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Text"
      >
        What type of products will be listed?
      </h2>
      <p
        class="components-truncate components-text css-sgni2d-View-Text-sx-Base em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Text"
      >
        Choose any that apply
      </p>
    </div>
    <div
      class="components-surface components-card css-1vyvcpq-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-card__body components-card-body css-1sfrl79-View-Body-borderRadius em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            class="components-base-control components-checkbox-control woocommerce-profile-wizard__checkbox css-wdf2ti-Wrapper e1puf3u4"
          >
            <div
              class="components-base-control__field css-igk9ll-StyledField e1puf3u3"
            >
              <span
                class="components-checkbox-control__input-container"
              >
                <input
                  class="components-checkbox-control__input"
                  id="inspector-checkbox-control-0"
                  type="checkbox"
                  value="1"
                />
              </span>
              <label
                class="components-checkbox-control__label"
                for="inspector-checkbox-control-0"
              >
                <span
                  class="woocommerce-product-wizard__product-types-label"
                >
                  Paid product
                </span>
                <button
                  aria-label="Learn more about recommended free business features"
                  class="components-button is-tertiary"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    focusable="false"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z"
                    />
                  </svg>
                </button>
                
                <span
                  class="components-truncate components-text woocommerce-pill css-1g4vivm-View-Text-sx-Base em57xhy0"
                  data-wp-c16t="true"
                  data-wp-component="Text"
                >
                  <span
                    class="screen-reader-text"
                  >
                    This product type requires a paid extension.
We'll add this to a cart so that
you can purchase and install it later.
                  </span>
                  $10 per month
                </span>
              </label>
            </div>
          </div>
          <div
            class="components-base-control components-checkbox-control woocommerce-profile-wizard__checkbox css-wdf2ti-Wrapper e1puf3u4"
          >
            <div
              class="components-base-control__field css-igk9ll-StyledField e1puf3u3"
            >
              <span
                class="components-checkbox-control__input-container"
              >
                <input
                  class="components-checkbox-control__input"
                  id="inspector-checkbox-control-1"
                  type="checkbox"
                  value="1"
                />
              </span>
              <label
                class="components-checkbox-control__label"
                for="inspector-checkbox-control-1"
              >
                Free product
              </label>
            </div>
          </div>
        </div>
        <div
          class="components-flex components-card__footer components-card-footer css-r62jpu-View-Flex-sx-Base-sx-Items-ItemsRow-Footer-borderRadius-borderColor-medium-borderless em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardFooter"
        >
          <button
            class="components-button is-primary"
            disabled=""
            type="button"
          >
            Continue
          </button>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
    <div
      class="woocommerce-profile-wizard__card-help-footnote"
    >
      <div
        class="woocommerce-profile-wizard__product-types-pricing-toggle woocommerce-profile-wizard__checkbox"
      >
        <label
          for="woocommerce-product-types__pricing-toggle"
        >
          <p
            class="components-truncate components-text css-sgni2d-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Display monthly prices
          </p>
          <span
            class="components-form-toggle is-checked"
          >
            <input
              checked=""
              class="components-form-toggle__input"
              id="woocommerce-product-types__pricing-toggle"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
        </label>
      </div>
      <span
        class="components-truncate components-text css-1g4vivm-View-Text-sx-Base em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Text"
      >
        Billing is annual. All purchases are covered by our 30 day money back guarantee and include access to support and updates. Extensions will be added to a cart for you to purchase later.
      </span>
    </div>
  </div>
</div>
`;

exports[`ProductTypes should show annual prices on toggle 1`] = `
<div>
  <div
    class="woocommerce-profile-wizard__product-types"
  >
    <div
      class="woocommerce-profile-wizard__step-header"
    >
      <h2
        class="components-truncate components-text css-7a2wz9-View-Text-sx-Base em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Text"
      >
        What type of products will be listed?
      </h2>
      <p
        class="components-truncate components-text css-sgni2d-View-Text-sx-Base em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Text"
      >
        Choose any that apply
      </p>
    </div>
    <div
      class="components-surface components-card css-1vyvcpq-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-card__body components-card-body css-1sfrl79-View-Body-borderRadius em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            class="components-base-control components-checkbox-control woocommerce-profile-wizard__checkbox css-wdf2ti-Wrapper e1puf3u4"
          >
            <div
              class="components-base-control__field css-igk9ll-StyledField e1puf3u3"
            >
              <span
                class="components-checkbox-control__input-container"
              >
                <input
                  class="components-checkbox-control__input"
                  id="inspector-checkbox-control-2"
                  type="checkbox"
                  value="1"
                />
              </span>
              <label
                class="components-checkbox-control__label"
                for="inspector-checkbox-control-2"
              >
                <span
                  class="woocommerce-product-wizard__product-types-label"
                >
                  Paid product
                </span>
                <button
                  aria-label="Learn more about recommended free business features"
                  class="components-button is-tertiary"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    focusable="false"
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z"
                    />
                  </svg>
                </button>
                
                <span
                  class="components-truncate components-text woocommerce-pill css-1g4vivm-View-Text-sx-Base em57xhy0"
                  data-wp-c16t="true"
                  data-wp-component="Text"
                >
                  <span
                    class="screen-reader-text"
                  >
                    This product type requires a paid extension.
We'll add this to a cart so that
you can purchase and install it later.
                  </span>
                  $120 per year
                </span>
              </label>
            </div>
          </div>
          <div
            class="components-base-control components-checkbox-control woocommerce-profile-wizard__checkbox css-wdf2ti-Wrapper e1puf3u4"
          >
            <div
              class="components-base-control__field css-igk9ll-StyledField e1puf3u3"
            >
              <span
                class="components-checkbox-control__input-container"
              >
                <input
                  class="components-checkbox-control__input"
                  id="inspector-checkbox-control-3"
                  type="checkbox"
                  value="1"
                />
              </span>
              <label
                class="components-checkbox-control__label"
                for="inspector-checkbox-control-3"
              >
                Free product
              </label>
            </div>
          </div>
        </div>
        <div
          class="components-flex components-card__footer components-card-footer css-r62jpu-View-Flex-sx-Base-sx-Items-ItemsRow-Footer-borderRadius-borderColor-medium-borderless em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardFooter"
        >
          <button
            class="components-button is-primary"
            disabled=""
            type="button"
          >
            Continue
          </button>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
    <div
      class="woocommerce-profile-wizard__card-help-footnote"
    >
      <div
        class="woocommerce-profile-wizard__product-types-pricing-toggle woocommerce-profile-wizard__checkbox"
      >
        <label
          for="woocommerce-product-types__pricing-toggle"
        >
          <p
            class="components-truncate components-text css-sgni2d-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Display monthly prices
          </p>
          <span
            class="components-form-toggle"
          >
            <input
              checked=""
              class="components-form-toggle__input"
              id="woocommerce-product-types__pricing-toggle"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
        </label>
      </div>
      <span
        class="components-truncate components-text css-1g4vivm-View-Text-sx-Base em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Text"
      >
        Billing is annual. All purchases are covered by our 30 day money back guarantee and include access to support and updates. Extensions will be added to a cart for you to purchase later.
      </span>
    </div>
  </div>
</div>
`;
